Applications/YOLO/PyTorch/main.py

```diff
@@ -0,0 +1,131 @@
+from yolo_loss import YoloV2_LOSS
+from dataset import YOLODataset, collate_db
+
+
+# set config
+out_size = 13
+num_classes = 5
+num_anchors = 5
+
+epochs = 1000
+batch_size = 8
+
+img_dir = './custom_dataset/images/*'
+ann_dir = './custom_dataset/annotations/*'
+
+
+# load data
+"""
+Need annotation text files corresponding to the name of the images.
+In the annotation, label for one object must be stored one line at a time,
+and data must be stored in the order of class number, x, y, w, h.
+
+example) 
+./images/abc.png
+./annotations/abc.txt
+    content of abc.txt: 2 20 20 40 40
+"""
+dataset = YOLODataset(img_dir, ann_dir)
+loader = DataLoader(dataset, batch_size=batch_size, collate_fn=collate_db, shuffle=True, drop_last=True)
+
+
+# set model, loss and optimizer
+model = YoloV2_light(num_classes=5)
+criterion = YoloV2_LOSS(num_classes=5)
+optimizer = optim.Adam(model.parameters(), lr=1e-3)
+scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=10, eta_min=0)
+
+
+# train model
+best_loss = 1e+10
+for epoch in range(epochs):
+    epoch_loss = 0
+    for idx, (img, bbox, cls) in enumerate(loader):
+        optimizer.zero_grad()
+        # model prediction
+        hypothesis = model(img).permute((0, 2, 3, 1))
+        hypothesis = hypothesis.reshape((batch_size, out_size**2, num_anchors, 5+num_classes))        
+        # split each prediction(bbox, iou, class prob)
+        bbox_pred_xy = torch.sigmoid(hypothesis[..., :2])
+        bbox_pred_wh = torch.exp(hypothesis[..., 2:4])
+        bbox_pred = torch.cat((bbox_pred_xy, bbox_pred_wh), 3)        
+        iou_pred = torch.sigmoid(hypothesis[..., 4:5])        
+        score_pred = hypothesis[..., 5:].contiguous()
+        prob_pred = torch.softmax(score_pred.view(-1, num_classes), dim=1).view(score_pred.shape)
+        # calc loss
+        loss = criterion(torch.FloatTensor(bbox_pred),
+                         torch.FloatTensor(iou_pred),
+                         torch.FloatTensor(prob_pred),
+                         bbox,
+                         cls)
+        # back prop
+        loss.backward()
+        optimizer.step()  
+        scheduler.step()
+        epoch_loss += loss.item()
+        
+    if epoch_loss < best_loss:
+        best_loss = epoch_loss
+        torch.save(model.state_dict(), './best_model.pt')
+        
+    print("{}epoch, loss: {:.4f}".format(epoch, epoch_loss / len(loader)))
+
+
+# post process for bbox
+def post_process_for_bbox(bbox_pred):
+    """
+    param bbox_pred: shape(batch_size, cell_h x cell_w, num_anchors, 4)
+    """
+    anchors = torch.FloatTensor(
+        [(1.3221, 1.73145),
+        (3.19275, 4.00944),
+        (5.05587, 8.09892),
+        (9.47112, 4.84053),
+        (11.2364, 10.0071)]
+    )
+
+    outsize = (13, 13)
+    width, height = outsize
+    
+    # restore cell pos to x, y    
+    for w in range(width):
+        for h in range(height):
+            bbox_pred[0, height*h + w, :, 0] += w
```

Comment:
I guess this code will only process first batch isn't it?

Reply:
You are right. thanks! I'll fix it.